lightning/src/routing/router.rs

```diff
@@ -222,78 +365,167 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
-								channel_features: $chan_features.clone(),
-								fee_msat: 0,
-								cltv_expiry_delta: 0,
-							},
-						)
-					});
-					if $src_node_id != *our_node_id {
-						// Ignore new_fee for channel-from-us as we assume all channels-from-us
-						// will have the same effective-fee
-						total_fee += new_fee;
-						if let Some(fee_inc) = final_value_msat.checked_add(total_fee).and_then(|inc| { (old_entry.2 as u64).checked_mul(inc) }) {
-							total_fee += fee_inc / 1000000 + (old_entry.1 as u64);
-						} else {
-							// max_value means we'll always fail the old_entry.0 > total_fee check
-							total_fee = u64::max_value();
-						}
+		// $following_hops_fees_msat represents the fees paid for using all the channel *after* this one,
+		// since that value has to be transferred over this channel.
+		( $chan_id: expr, $src_node_id: expr, $dest_node_id: expr, $directional_info: expr, $capacity_sats: expr, $chan_features: expr, $following_hops_fees_msat: expr ) => {
+			let available_liquidity_msat = bookkeeped_channels_liquidity_available_msat.entry($chan_id.clone()).or_insert_with(|| {
+				let mut initial_liquidity_available_msat = None;
+				if let Some(capacity_sats) = $capacity_sats {
+					initial_liquidity_available_msat = Some(capacity_sats * 1000);
+				}
+
+				if let Some(htlc_maximum_msat) = $directional_info.htlc_maximum_msat {
+					if let Some(available_msat) = initial_liquidity_available_msat {
+						initial_liquidity_available_msat = Some(cmp::min(available_msat, htlc_maximum_msat));
+					} else {
+						initial_liquidity_available_msat = Some(htlc_maximum_msat);
 					}
-					let new_graph_node = RouteGraphNode {
-						pubkey: $src_node_id,
-						lowest_fee_to_peer_through_node: total_fee,
-						lowest_fee_to_node: $starting_fee_msat as u64 + new_fee,
+				}
+
+				match initial_liquidity_available_msat {
+					Some(available_msat) => available_msat,
+					// We assume channels with unknown balance have a capacity of 0.0001 BTC (or 10_000 sats).
+					None => 10_000 * 1000
+				}
+			});
+
+			// Routing Fragmentation Mitigation heuristic:
+			//
+			// Routing fragmentation across many payment paths increases the overall routing
+			// fees as you have irreducible routing fees per-link used (`fee_base_msat`).
+			// Taking too many paths also smaller paths also increases the chance of payment failure.
+			// Thus to avoid this effect, we require from our collected links to provide
+			// at least a minimal liquidity contribution to the recommended value yet-to-be-fulfilled.
+			//
+			// This requirement is currently 5% of the already-collected value. This means as
+			// we successfully advance in our collection, the absolute liquidity contribution is lowered,
+			// thus increasing the number of potential channels to be selected.
+
+			// Update the absolute liquidity left to collect from previously built paths.
+			let value_left_to_collect_msat = recommended_value_msat - already_collected_value_msat;
```

Comment:
Well, the extra value can be striped at the end of the function anyway.
I think both of the ways achieve the goal, but perhaps your solution finds paths *better* (but not over-provisioning the very first path). Anyway, since it also passes the tests, I'm taking your suggestion.